File: order/consumer.py
import pika
import sys
import os
import json

QUEUE_NAME = "queue1"
EXCHANGE_NAME = "queue1.exchange"
DLX_NAME = "queue1.dlx"
RETRY_QUEUE_NAME = "queue1.retry"
DEAD_QUEUE_NAME = "queue1.dead"

MAX_RETRIES = 5
RETRY_TTL_MS = 30000  # 30s backoff between retries


def get_death_count(properties, queue_name):
    headers = properties.headers or {}
    x_death = headers.get("x-death")
    if not x_death:
        return 0
    for entry in x_death:
        if entry.get("queue") == queue_name:
            return entry.get("count", 0)
    return 0


def main():
    connection = pika.BlockingConnection(
        pika.ConnectionParameters(host=os.environ.get("RABBITMQ_HOST", "localhost"))
    )
    channel = connection.channel()

    channel.exchange_declare(exchange=EXCHANGE_NAME, exchange_type="direct", durable=True)
    channel.exchange_declare(exchange=DLX_NAME, exchange_type="direct", durable=True)

    # Main queue, dead-letters into the retry flow on nack
    channel.queue_declare(
        queue=QUEUE_NAME,
        durable=True,
        arguments={
            "x-dead-letter-exchange": DLX_NAME,
            "x-dead-letter-routing-key": "queue1.retry",
        },
    )
    channel.queue_bind(queue=QUEUE_NAME, exchange=EXCHANGE_NAME, routing_key=QUEUE_NAME)

    # Retry queue: holds message for RETRY_TTL_MS, then dead-letters back to the main queue
    channel.queue_declare(
        queue=RETRY_QUEUE_NAME,
        durable=True,
        arguments={
            "x-message-ttl": RETRY_TTL_MS,
            "x-dead-letter-exchange": EXCHANGE_NAME,
            "x-dead-letter-routing-key": QUEUE_NAME,
        },
    )
    channel.queue_bind(queue=RETRY_QUEUE_NAME, exchange=DLX_NAME, routing_key="queue1.retry")

    # Terminal dead queue: no further DLX, poison messages land here for good
    channel.queue_declare(queue=DEAD_QUEUE_NAME, durable=True)
    channel.queue_bind(queue=DEAD_QUEUE_NAME, exchange=DLX_NAME, routing_key="queue1.dead")

    def callback(ch, method, properties, body):
        try:
            message = body.decode("utf-8")
            message_dict = json.loads(message)
            print("[x] received %r" % message_dict)

            # TODO: do the actual processing here (e.g. call the order API).
            # If that call can fail, wrap it in its own try/except below so
            # parse errors and processing errors are told apart.

            ch.basic_ack(delivery_tag=method.delivery_tag)

        except json.JSONDecodeError as parse_error:
            # Malformed JSON will never parse successfully no matter how many
            # times we retry, so send straight to the dead queue.
            print(f"Error parsing message, routing to dead queue: {parse_error}")
            ch.basic_publish(
                exchange=DLX_NAME,
                routing_key="queue1.dead",
                body=body,
                properties=properties,
            )
            ch.basic_ack(delivery_tag=method.delivery_tag)

        except Exception as error:
            death_count = get_death_count(properties, QUEUE_NAME)
            print(f"Error processing message (attempt {death_count + 1}): {error}")

            if death_count >= MAX_RETRIES:
                ch.basic_publish(
                    exchange=DLX_NAME,
                    routing_key="queue1.dead",
                    body=body,
                    properties=properties,
                )
                ch.basic_ack(delivery_tag=method.delivery_tag)
            else:
                # requeue=False -> DLX -> retry queue -> back to main queue after TTL
                ch.basic_nack(delivery_tag=method.delivery_tag, requeue=False)

    channel.basic_qos(prefetch_count=1)
    channel.basic_consume(queue=QUEUE_NAME, on_message_callback=callback, auto_ack=False)

    print(" [*] waiting for the messages. To exit press Ctrl-C")
    channel.start_consuming()


if __name__ == "__main__":
    try:
        main()
    except KeyboardInterrupt:
        print("Interrupted")
        try:
            sys.exit(0)
        except SystemExit:
            os._exit(0)

File: order_processor_node/consumer.js
const amqp = require('amqplib');
const axios = require('axios');
const dotenv = require('dotenv');
dotenv.config();

const order_api_url = process.env.ORDER_API_URL || 'http://localhost:5002';
const queueName = 'queue1';
const exchangeName = 'queue1.exchange';
const dlxName = 'queue1.dlx';
const retryQueueName = 'queue1.retry';
const deadQueueName = 'queue1.dead';
const rabbitMQHost = process.env.RABBITMQ_HOST || 'amqp://localhost';

const MAX_RETRIES = 5;
const RETRY_TTL_MS = 30000; // 30s backoff between retries

function getDeathCount(msg) {
  const xDeath = msg.properties?.headers?.['x-death'];
  if (!xDeath) return 0;
  const entry = xDeath.find((d) => d.queue === queueName);
  return entry?.count ?? 0;
}

async function consumeMessages() {
  console.log('consumeMessages');
  const connection = await amqp.connect(rabbitMQHost);
  const channel = await connection.createChannel();

  await channel.assertExchange(exchangeName, 'direct', { durable: true });
  await channel.assertExchange(dlxName, 'direct', { durable: true });

  // Main queue, dead-letters into the retry flow on nack
  await channel.assertQueue(queueName, {
    durable: true,
    arguments: {
      'x-dead-letter-exchange': dlxName,
      'x-dead-letter-routing-key': 'queue1.retry',
    },
  });
  await channel.bindQueue(queueName, exchangeName, queueName);

  // Retry queue: holds message for RETRY_TTL_MS, then dead-letters back to the main queue
  await channel.assertQueue(retryQueueName, {
    durable: true,
    arguments: {
      'x-message-ttl': RETRY_TTL_MS,
      'x-dead-letter-exchange': exchangeName,
      'x-dead-letter-routing-key': queueName,
    },
  });
  await channel.bindQueue(retryQueueName, dlxName, 'queue1.retry');

  // Terminal dead queue: no further DLX, poison messages land here for good
  await channel.assertQueue(deadQueueName, { durable: true });
  await channel.bindQueue(deadQueueName, dlxName, 'queue1.dead');

  console.log(`[*] Waiting for messages in ${queueName}. To exit, press Ctrl-C`);

  channel.consume(queueName, async (msg) => {
    if (msg === null) return;

    let messageObject;
    try {
      const message = msg.content.toString();
      messageObject = JSON.parse(message);
      console.log('[x] Received', messageObject);
    } catch (parseError) {
      // Malformed JSON will never parse successfully no matter how many
      // times we retry, so send straight to the dead queue.
      console.error('Error parsing message, routing to dead queue:', parseError);
      channel.publish(dlxName, 'queue1.dead', msg.content, { headers: msg.properties.headers });
      channel.ack(msg);
      return;
    }

    try {
      messageObject.is_open = false;
      const putUrl = `${order_api_url}/${messageObject.id}`;
      console.log(putUrl, messageObject);
      const response = await axios.put(putUrl, messageObject);
      console.log('Response:', response.data);
      channel.ack(msg); // only ack after the PUT actually succeeds
    } catch (error) {
      const deathCount = getDeathCount(msg);
      console.error(`Error sending PUT request (attempt ${deathCount + 1}):`, error.message);

      if (deathCount >= MAX_RETRIES) {
        channel.publish(dlxName, 'queue1.dead', msg.content, { headers: msg.properties.headers });
        channel.ack(msg);
      } else {
        channel.nack(msg, false, false); // -> DLX -> retry queue -> back to main queue after TTL
      }
    }
  });
}

module.exports = consumeMessages;
